Multiple "Unmarshal" calls on the same data slice lead to corrupted data that cannot be unmarshaled

We (@symflower) are currently in the process of migrating to a faster JSON package than "encoding/json" and this project seems to be a good replacement. However, before we can benchmark any further we need to get rid of (hopefully only) this bug.

This reproducer works with "encoding/json" but not with "github.com/goccy/go-json". I tried to minimize the problem down to how we handle dynamic types with JSON. We basically have multiple "Unmarshal" calls that look at the types and then execute the right unmarshal functionality with the given type. Minimized only the calls to "UNmarshal" truly matter.

```
package main

import (
	"fmt"

	gojson "github.com/goccy/go-json"
)

func main() {
	data := []byte(`[
		{
		  "Body": {
			"List": [
			  {
				"nodeType": "Call"
			  },
			  {
				"nodeType": "Call"
			  }
			],
			"nodeKind": "Block",
			"nodeType": "Statement"
		  },
		  "nodeType": "Function"
		},
		{
		  "Body": {
			"List": [],
			"nodeKind": "Block",
			"nodeType": "Statement"
		  },
		  "nodeType": "Function"
		}
	  ]
	`)

	if !gojson.Valid(data) {
		panic("stop")
	}

	var a []gojson.RawMessage
	if err := gojson.Unmarshal(data, &a); err != nil {
		panic(err)
	}

	fmt.Println("Original:")
	for i, data := range a {
		fmt.Printf("%d: %s\n", i, data)
	}

	fmt.Println("when iterated while unmarshaling:")
	for i, data := range a {
		fmt.Printf("%d: %s\n", i, data)

		var b map[string]gojson.RawMessage
		if err := gojson.Unmarshal(data, &b); err != nil {
			panic(err)
		}

		var c map[string]gojson.RawMessage
		if err := gojson.Unmarshal(b["Body"], &c); err != nil {
			panic(err)
		}

		var d []gojson.RawMessage
		if err := gojson.Unmarshal(c["List"], &d); err != nil {
			panic(err)
		}
	}
}

```

Produces:

```
Original:
0: {
                  "Body": {
                        "List": [
                          {
                                "nodeType": "Call"
                          },
                          {
                                "nodeType": "Call"
                          }
                        ],
                        "nodeKind": "Block",
                        "nodeType": "Statement"
                  },
                  "nodeType": "Function"
                }
1: {
                  "Body": {
                        "List": [],
                        "nodeKind": "Block",
                        "nodeType": "Statement"
                  },
                  "nodeType": "Function"
                }
when iterated while unmarshaling:
0: {
                  "Body": {
                        "List": [
                          {
                                "nodeType": "Call"
                          },
                          {
                                "nodeType": "Call"
                          }
                        ],
                        "nodeKind": "Block",
                        "nodeType": "Statement"
                  },
                  "nodeType": "Function"
                }
1: {
                                "nodeType": "Call"
                          }                     "nodeKind": "Block",
                        "nodeType": "Statement"
                  },
                  "nodeType": "Function"
                }
panic: invalid character '"' after top-level value

goroutine 1 [running]:
main.main()
        /home/zimmski/symflower/tt.go:59 +0x6c5
exit status 2
```

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -121,17 +121,16 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
 				if d.isElemUnmarshalJSONType {
 					receiver := unsafe_New(d.elemType)
-					if err := d.valueDecoder.decodeStream(s, depth, receiver); err != nil {
-						return err
+					if d.elemType.Kind() == reflect.Slice {
+						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = receiver
+					} else {
+						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
 					}
-					*(*unsafe.Pointer)(ep) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
-				} else {
-					if d.isElemPointerType {
-						*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
-					}
-					if err := d.valueDecoder.decodeStream(s, depth, ep); err != nil {
-						return err
-					}
+				} else if d.isElemPointerType {
+					**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = nil // initialize elem pointer
+				}
+				if err := d.valueDecoder.decodeStream(s, depth, ep); err != nil {
+					return err
 				}
 				s.skipWhiteSpace()
 			RETRY:
@@ -244,25 +243,18 @@ func (d *sliceDecoder) decode(buf []byte, cursor, depth int64, p unsafe.Pointer)
 				if d.isElemUnmarshalJSONType {
 					receiver := unsafe_New(d.elemType)
 					if d.elemType.Kind() == reflect.Slice {
-						*(*unsafe.Pointer)(ep) = receiver
+						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = receiver
 					} else {
-						*(*unsafe.Pointer)(ep) = *(*unsafe.Pointer)(receiver)
+						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
 					}
-					c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
-					if err != nil {
-						return 0, err
-					}
-					cursor = c
-				} else {
-					if d.isElemPointerType {
-						*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
-					}
-					c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
-					if err != nil {
-						return 0, err
-					}
-					cursor = c
+				} else if d.isElemPointerType {
+					**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = nil // initialize elem pointer
 				}
+				c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
+				if err != nil {
+					return 0, err
+				}
+				cursor = c
 				cursor = skipWhiteSpace(buf, cursor)
 				switch buf[cursor] {
 				case ']':

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -67,9 +67,6 @@ func copySlice(elemType *rtype, dst, src sliceHeader) int
 //go:linkname newArray reflect.unsafe_NewArray
 func newArray(*rtype, int) unsafe.Pointer
 
-//go:linkname typedmemmovepartial reflect.typedmemmovepartial
-func typedmemmovepartial(typ *rtype, dst, src unsafe.Pointer, off, size uintptr)
-
 //go:linkname typedmemmove reflect.typedmemmove
 func typedmemmove(t *rtype, dst, src unsafe.Pointer)
 

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -9,6 +9,7 @@ import (
 type sliceDecoder struct {
 	elemType          *rtype
 	isElemPointerType bool
+	isElemSliceType   bool
 	valueDecoder      decoder
 	size              uintptr
 	arrayPool         sync.Pool
@@ -34,6 +35,7 @@ func newSliceDecoder(dec decoder, elemType *rtype, size uintptr, structName, fie
 		valueDecoder:      dec,
 		elemType:          elemType,
 		isElemPointerType: elemType.Kind() == reflect.Ptr || elemType.Kind() == reflect.Map,
+		isElemSliceType:   elemType.Kind() == reflect.Slice,
 		size:              size,
 		arrayPool: sync.Pool{
 			New: func() interface{} {
@@ -233,6 +235,12 @@ func (d *sliceDecoder) decode(buf []byte, cursor, depth int64, p unsafe.Pointer)
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
 				if d.isElemPointerType {
 					*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
+				} else if d.isElemSliceType {
+					*(*sliceHeader)(ep) = sliceHeader{
+						data: newArray(d.elemType, 0),
+						len:  0,
+						cap:  0,
+					}
 				}
 				c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
 				if err != nil {

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -121,6 +121,12 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
 				if d.isElemPointerType {
 					*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
+				} else if d.isElemSliceType {
+					*(*sliceHeader)(ep) = sliceHeader{
+						data: newArray(d.elemType, 0),
+						len:  0,
+						cap:  0,
+					}
 				}
 				if err := d.valueDecoder.decodeStream(s, depth, ep); err != nil {
 					return err

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -67,6 +67,12 @@ func copySlice(elemType *rtype, dst, src sliceHeader) int
 //go:linkname newArray reflect.unsafe_NewArray
 func newArray(*rtype, int) unsafe.Pointer
 
+//go:linkname typedmemmovepartial reflect.typedmemmovepartial
+func typedmemmovepartial(typ *rtype, dst, src unsafe.Pointer, off, size uintptr)
+
+//go:linkname typedmemmove reflect.typedmemmove
+func typedmemmove(t *rtype, dst, src unsafe.Pointer)
+
 func (d *sliceDecoder) errNumber(offset int64) *UnmarshalTypeError {
 	return &UnmarshalTypeError{
 		Value:  "number",
@@ -120,12 +126,8 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
 				}
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
 				if d.isElemUnmarshalJSONType {
-					receiver := unsafe_New(d.elemType)
-					if d.elemType.Kind() == reflect.Slice {
-						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = receiver
-					} else {
-						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
-					}
+					// assign new element to the slice
+					typedmemmove(d.elemType, ep, unsafe_New(d.elemType))
 				} else if d.isElemPointerType {
 					**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = nil // initialize elem pointer
 				}
@@ -241,12 +243,8 @@ func (d *sliceDecoder) decode(buf []byte, cursor, depth int64, p unsafe.Pointer)
 				}
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
 				if d.isElemUnmarshalJSONType {
-					receiver := unsafe_New(d.elemType)
-					if d.elemType.Kind() == reflect.Slice {
-						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = receiver
-					} else {
-						**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
-					}
+					// assign new element to the slice
+					typedmemmove(d.elemType, ep, unsafe_New(d.elemType))
 				} else if d.isElemPointerType {
 					**(**unsafe.Pointer)(unsafe.Pointer(&ep)) = nil // initialize elem pointer
 				}

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -7,14 +7,14 @@ import (
 )
 
 type sliceDecoder struct {
-	elemType          *rtype
-	isElemPointerType bool
-	isElemSliceType   bool
-	valueDecoder      decoder
-	size              uintptr
-	arrayPool         sync.Pool
-	structName        string
-	fieldName         string
+	elemType                *rtype
+	isElemPointerType       bool
+	isElemUnmarshalJSONType bool
+	valueDecoder            decoder
+	size                    uintptr
+	arrayPool               sync.Pool
+	structName              string
+	fieldName               string
 }
 
 // If use reflect.SliceHeader, data type is uintptr.
@@ -32,11 +32,11 @@ const (
 
 func newSliceDecoder(dec decoder, elemType *rtype, size uintptr, structName, fieldName string) *sliceDecoder {
 	return &sliceDecoder{
-		valueDecoder:      dec,
-		elemType:          elemType,
-		isElemPointerType: elemType.Kind() == reflect.Ptr || elemType.Kind() == reflect.Map,
-		isElemSliceType:   elemType.Kind() == reflect.Slice,
-		size:              size,
+		valueDecoder:            dec,
+		elemType:                elemType,
+		isElemPointerType:       elemType.Kind() == reflect.Ptr || elemType.Kind() == reflect.Map,
+		isElemUnmarshalJSONType: rtype_ptrTo(elemType).Implements(unmarshalJSONType),
+		size:                    size,
 		arrayPool: sync.Pool{
 			New: func() interface{} {
 				return &sliceHeader{
@@ -119,17 +119,19 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
 					copySlice(d.elemType, dst, src)
 				}
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
-				if d.isElemPointerType {
-					*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
-				} else if d.isElemSliceType {
-					*(*sliceHeader)(ep) = sliceHeader{
-						data: newArray(d.elemType, 0),
-						len:  0,
-						cap:  0,
+				if d.isElemUnmarshalJSONType {
+					receiver := unsafe_New(d.elemType)
+					if err := d.valueDecoder.decodeStream(s, depth, receiver); err != nil {
+						return err
 					}
-				}
-				if err := d.valueDecoder.decodeStream(s, depth, ep); err != nil {
-					return err
+					*(*unsafe.Pointer)(ep) = **(**unsafe.Pointer)(unsafe.Pointer(&receiver))
+				} else {
+					if d.isElemPointerType {
+						*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
+					}
+					if err := d.valueDecoder.decodeStream(s, depth, ep); err != nil {
+						return err
+					}
 				}
 				s.skipWhiteSpace()
 			RETRY:
@@ -239,20 +241,28 @@ func (d *sliceDecoder) decode(buf []byte, cursor, depth int64, p unsafe.Pointer)
 					copySlice(d.elemType, dst, src)
 				}
 				ep := unsafe.Pointer(uintptr(data) + uintptr(idx)*d.size)
-				if d.isElemPointerType {
-					*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
-				} else if d.isElemSliceType {
-					*(*sliceHeader)(ep) = sliceHeader{
-						data: newArray(d.elemType, 0),
-						len:  0,
-						cap:  0,
+				if d.isElemUnmarshalJSONType {
+					receiver := unsafe_New(d.elemType)
+					if d.elemType.Kind() == reflect.Slice {
+						*(*unsafe.Pointer)(ep) = receiver
+					} else {
+						*(*unsafe.Pointer)(ep) = *(*unsafe.Pointer)(receiver)
 					}
+					c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
+					if err != nil {
+						return 0, err
+					}
+					cursor = c
+				} else {
+					if d.isElemPointerType {
+						*(*unsafe.Pointer)(ep) = nil // initialize elem pointer
+					}
+					c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
+					if err != nil {
+						return 0, err
+					}
+					cursor = c
 				}
-				c, err := d.valueDecoder.decode(buf, cursor, depth, ep)
-				if err != nil {
-					return 0, err
-				}
-				cursor = c
 				cursor = skipWhiteSpace(buf, cursor)
 				switch buf[cursor] {
 				case ']':

--- a/decode_test.go
+++ b/decode_test.go
@@ -2996,6 +2996,53 @@ func TestDecodeMultipleUnmarshal(t *testing.T) {
 	})
 }
 
+func TestMultipleDecodeWithRawMessage(t *testing.T) {
+	original := []byte(`{
+		"Body": {
+			"List": [
+				{
+					"Returns": [
+						{
+							"Value": "10",
+							"nodeType": "Literal"
+						}
+					],
+					"nodeKind": "Return",
+					"nodeType": "Statement"
+				}
+			],
+			"nodeKind": "Block",
+			"nodeType": "Statement"
+		},
+		"nodeType": "Function"
+	}`)
+
+	var a map[string]json.RawMessage
+	if err := json.Unmarshal(original, &a); err != nil {
+		t.Fatal(err)
+	}
+	var b map[string]json.RawMessage
+	if err := json.Unmarshal(a["Body"], &b); err != nil {
+		t.Fatal(err)
+	}
+	var c []json.RawMessage
+	if err := json.Unmarshal(b["List"], &c); err != nil {
+		t.Fatal(err)
+	}
+	var d map[string]json.RawMessage
+	if err := json.Unmarshal(c[0], &d); err != nil {
+		t.Fatal(err)
+	}
+	var e []json.RawMessage
+	if err := json.Unmarshal(d["Returns"], &e); err != nil {
+		t.Fatal(err)
+	}
+	var f map[string]json.RawMessage
+	if err := json.Unmarshal(e[0], &f); err != nil {
+		t.Fatal(err)
+	}
+}
+
 type intUnmarshaler int
 
 func (u *intUnmarshaler) UnmarshalJSON(b []byte) error {

--- a/decode_test.go
+++ b/decode_test.go
@@ -2931,6 +2931,71 @@ func TestDecodeSlice(t *testing.T) {
 	}
 }
 
+func TestDecodeMultipleUnmarshal(t *testing.T) {
+	data := []byte(`[{"AA":{"X":[{"a": "A"},{"b": "B"}],"Y":"y","Z":"z"},"BB":"bb"},{"AA":{"X":[],"Y":"y","Z":"z"},"BB":"bb"}]`)
+	var a []json.RawMessage
+	if err := json.Unmarshal(data, &a); err != nil {
+		t.Fatal(err)
+	}
+	if len(a) != 2 {
+		t.Fatalf("failed to decode: got %v", a)
+	}
+	t.Run("first", func(t *testing.T) {
+		data := a[0]
+		var v map[string]json.RawMessage
+		if err := json.Unmarshal(data, &v); err != nil {
+			t.Fatal(err)
+		}
+		if string(v["AA"]) != `{"X":[{"a": "A"},{"b": "B"}],"Y":"y","Z":"z"}` {
+			t.Fatalf("failed to decode. got %q", v["AA"])
+		}
+		var aa map[string]json.RawMessage
+		if err := json.Unmarshal(v["AA"], &aa); err != nil {
+			t.Fatal(err)
+		}
+		if string(aa["X"]) != `[{"a": "A"},{"b": "B"}]` {
+			t.Fatalf("failed to decode. got %q", v["X"])
+		}
+		var x []json.RawMessage
+		if err := json.Unmarshal(aa["X"], &x); err != nil {
+			t.Fatal(err)
+		}
+		if len(x) != 2 {
+			t.Fatalf("failed to decode: %v", x)
+		}
+		if string(x[0]) != `{"a": "A"}` {
+			t.Fatal("failed to decode")
+		}
+		if string(x[1]) != `{"b": "B"}` {
+			t.Fatal("failed to decode")
+		}
+	})
+	t.Run("second", func(t *testing.T) {
+		data := a[1]
+		var v map[string]json.RawMessage
+		if err := json.Unmarshal(data, &v); err != nil {
+			t.Fatal(err)
+		}
+		if string(v["AA"]) != `{"X":[],"Y":"y","Z":"z"}` {
+			t.Fatalf("failed to decode. got %q", v["AA"])
+		}
+		var aa map[string]json.RawMessage
+		if err := json.Unmarshal(v["AA"], &aa); err != nil {
+			t.Fatal(err)
+		}
+		if string(aa["X"]) != `[]` {
+			t.Fatalf("failed to decode. got %q", v["X"])
+		}
+		var x []json.RawMessage
+		if err := json.Unmarshal(aa["X"], &x); err != nil {
+			t.Fatal(err)
+		}
+		if len(x) != 0 {
+			t.Fatalf("failed to decode: %v", x)
+		}
+	})
+}
+
 func TestInvalidTopLevelValue(t *testing.T) {
 	t.Run("invalid end of buffer", func(t *testing.T) {
 		var v struct{}

--- a/decode_test.go
+++ b/decode_test.go
@@ -2996,6 +2996,156 @@ func TestDecodeMultipleUnmarshal(t *testing.T) {
 	})
 }
 
+type intUnmarshaler int
+
+func (u *intUnmarshaler) UnmarshalJSON(b []byte) error {
+	if *u != 0 {
+		return fmt.Errorf("failed to decode of slice with int unmarshaler")
+	}
+	*u = 10
+	return nil
+}
+
+type arrayUnmarshaler [5]int
+
+func (u *arrayUnmarshaler) UnmarshalJSON(b []byte) error {
+	if (*u)[0] != 0 {
+		return fmt.Errorf("failed to decode of slice with array unmarshaler")
+	}
+	(*u)[0] = 10
+	return nil
+}
+
+type mapUnmarshaler map[string]int
+
+func (u *mapUnmarshaler) UnmarshalJSON(b []byte) error {
+	if len(*u) != 0 {
+		return fmt.Errorf("failed to decode of slice with map unmarshaler")
+	}
+	*u = map[string]int{"a": 10}
+	return nil
+}
+
+type structUnmarshaler struct {
+	A int
+}
+
+func (u *structUnmarshaler) UnmarshalJSON(b []byte) error {
+	if u.A != 0 {
+		return fmt.Errorf("failed to decode of slice with struct unmarshaler")
+	}
+	u.A = 10
+	return nil
+}
+
+func TestSliceElemUnmarshaler(t *testing.T) {
+	t.Run("int", func(t *testing.T) {
+		var v []intUnmarshaler
+		if err := json.Unmarshal([]byte(`[1,2,3,4,5]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 5 {
+			t.Fatalf("failed to decode of slice with int unmarshaler: %v", v)
+		}
+		if v[0] != 10 {
+			t.Fatalf("failed to decode of slice with int unmarshaler: %v", v)
+		}
+		if err := json.Unmarshal([]byte(`[6]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 1 {
+			t.Fatalf("failed to decode of slice with int unmarshaler: %v", v)
+		}
+		if v[0] != 10 {
+			t.Fatalf("failed to decode of slice with int unmarshaler: %v", v)
+		}
+	})
+	t.Run("slice", func(t *testing.T) {
+		var v []json.RawMessage
+		if err := json.Unmarshal([]byte(`[1,2,3,4,5]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 5 {
+			t.Fatalf("failed to decode of slice with slice unmarshaler: %v", v)
+		}
+		if len(v[0]) != 1 {
+			t.Fatalf("failed to decode of slice with slice unmarshaler: %v", v)
+		}
+		if err := json.Unmarshal([]byte(`[6]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 1 {
+			t.Fatalf("failed to decode of slice with slice unmarshaler: %v", v)
+		}
+		if len(v[0]) != 1 {
+			t.Fatalf("failed to decode of slice with slice unmarshaler: %v", v)
+		}
+	})
+	t.Run("array", func(t *testing.T) {
+		var v []arrayUnmarshaler
+		if err := json.Unmarshal([]byte(`[1,2,3,4,5]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 5 {
+			t.Fatalf("failed to decode of slice with array unmarshaler: %v", v)
+		}
+		if v[0][0] != 10 {
+			t.Fatalf("failed to decode of slice with array unmarshaler: %v", v)
+		}
+		if err := json.Unmarshal([]byte(`[6]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 1 {
+			t.Fatalf("failed to decode of slice with array unmarshaler: %v", v)
+		}
+		if v[0][0] != 10 {
+			t.Fatalf("failed to decode of slice with array unmarshaler: %v", v)
+		}
+	})
+	t.Run("map", func(t *testing.T) {
+		var v []mapUnmarshaler
+		if err := json.Unmarshal([]byte(`[{"a":1},{"b":2},{"c":3},{"d":4},{"e":5}]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 5 {
+			t.Fatalf("failed to decode of slice with map unmarshaler: %v", v)
+		}
+		if v[0]["a"] != 10 {
+			t.Fatalf("failed to decode of slice with map unmarshaler: %v", v)
+		}
+		if err := json.Unmarshal([]byte(`[6]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 1 {
+			t.Fatalf("failed to decode of slice with map unmarshaler: %v", v)
+		}
+		if v[0]["a"] != 10 {
+			t.Fatalf("failed to decode of slice with map unmarshaler: %v", v)
+		}
+	})
+	t.Run("struct", func(t *testing.T) {
+		var v []structUnmarshaler
+		if err := json.Unmarshal([]byte(`[1,2,3,4,5]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 5 {
+			t.Fatalf("failed to decode of slice with struct unmarshaler: %v", v)
+		}
+		if v[0].A != 10 {
+			t.Fatalf("failed to decode of slice with struct unmarshaler: %v", v)
+		}
+		if err := json.Unmarshal([]byte(`[6]`), &v); err != nil {
+			t.Fatal(err)
+		}
+		if len(v) != 1 {
+			t.Fatalf("failed to decode of slice with struct unmarshaler: %v", v)
+		}
+		if v[0].A != 10 {
+			t.Fatalf("failed to decode of slice with struct unmarshaler: %v", v)
+		}
+	})
+}
+
 func TestInvalidTopLevelValue(t *testing.T) {
 	t.Run("invalid end of buffer", func(t *testing.T) {
 		var v struct{}

